File: frontend/src/engine/capabilities/autoFitRows.ts
/**
 * autoFitRows — row-axis pair of autoFitColumns.
 *
 * Office.js: Range.getEntireRow().format.autoFitRows() (ExcelApi 1.2+).
 */

import { CapabilityMeta, AutoFitRowsParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "autoFitRows",
  description: "Auto-fit row heights — row-axis pair of autoFitColumns",
  mutates: false,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.2",
};

async function handler(
  context: Excel.RequestContext,
  params: AutoFitRowsParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { range: address, sheetName } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would auto-fit rows in ${address ?? "active sheet used range"}` };
  }

  options.onProgress?.("Auto-fitting row heights...");

  let target: Excel.Range;
  if (address) {
    target = resolveRange(context, address);
  } else {
    const sheet = sheetName
      ? context.workbook.worksheets.getItem(sheetName)
      : context.workbook.worksheets.getActiveWorksheet();
    target = sheet.getUsedRange(true);
  }
  target.getEntireRow().format.autofitRows();
  target.load("address");
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Auto-fit rows on ${target.address}.`,
    outputs: { range: target.address },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/auto_fit_rows.py
"""autoFitRows — auto-fit row heights on a range (or the sheet's used range)."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    sheet_name = params.get("sheetName")

    if address:
        rng = resolve_range(ctx.workbook_handle, address)
    else:
        book = ctx.workbook_handle
        sheet = book.sheets[sheet_name] if sheet_name else book.sheets.active
        rng = sheet.used_range

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would autofit rows on {rng.address}."}

    try:
        rng.rows.autofit()
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"AutoFit rows failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Autofit rows on {rng.address}.",
        "outputs": {"range": rng.address},
    }


registry.register("autoFitRows", handler, mutates=False, affects_formatting=True)
